kms.go

```diff
@@ -0,0 +1,186 @@
+package hdfs
+
+import (
+	"bytes"
+	"crypto/aes"
+	"io/ioutil"
+
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	hdfs "github.com/colinmarc/hdfs/v2/internal/protocol/hadoop_hdfs"
+	spnego "github.com/jcmturner/gokrb5/v8/spnego"
+)
+
+const (
+	kmsSchemeHTTP  = "kms://http@"
+	kmsSchemeHTTPS = "kms://https@"
+)
+
+func (c *Client) kmsAuth(url string) error {
+	if c.options.KerberosServicePrincipleName == "" {
+		url += ("&user.name=" + c.options.User)
+	}
+
+	req, err := http.NewRequest("OPTIONS", url, nil)
+	if err != nil {
+		return err
+	}
+
+	var resp *http.Response
+	if c.options.KerberosClient != nil {
+		kHttp := spnego.NewClient(c.options.KerberosClient, c.http, "")
+		resp, err = kHttp.Do(req)
+	} else {
+		resp, err = c.http.Do(req)
+	}
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		return fmt.Errorf("bad kms auth status: %v", resp.StatusCode)
+	}
+	return nil
+}
+
+func kmsSplit(keyProvider string) []string {
```

Comment:
no need for a separate function here

Reply:
i've improved kms address parsing, and this function got replaced by `kmsParseProviderUri`.